patches/server/0725-OfflinePlayer-advancements.patch

```diff
@@ -0,0 +1,64 @@
+From: Jake Potrebic <jake.m.potrebic@gmail.com>
+Date: Thu, 22 Apr 2021 01:41:56 -0700
+Subject: [PATCH] OfflinePlayer advancements
+
+
+diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
+index 65657c009f6d5a5d5740e80f912a5893333c7085..8a5d76d8f26c292fde93f3731442ac145e20653f 100644
+--- a/src/main/java/net/minecraft/server/players/PlayerList.java
++++ b/src/main/java/net/minecraft/server/players/PlayerList.java
+@@ -1463,6 +1463,20 @@ public abstract class PlayerList {
+         advancementdataplayer.setPlayer(player);
+         return advancementdataplayer;
+     }
++    // Paper start
++    public PlayerAdvancements getPlayerAdvancements(UUID uuid) {
++        ServerPlayer entityPlayer = getPlayer(uuid);
++        PlayerAdvancements advancementDataPlayer = entityPlayer == null ? null : entityPlayer.getAdvancements();
++        if (advancementDataPlayer == null) {
++            File file = this.server.getWorldPath(LevelResource.PLAYER_ADVANCEMENTS_DIR).toFile();
++            File file1 = new File(file, uuid + ".json");
++
++            advancementDataPlayer = new PlayerAdvancements(this.server.getFixerUpper(), this, this.server.getAdvancements(), file1, entityPlayer);
++        }
++        advancementDataPlayer.setPlayer(entityPlayer);
++        return advancementDataPlayer;
++    }
++    // Paper end
+ 
+     public void setViewDistance(int viewDistance) {
+         this.viewDistance = viewDistance;
+diff --git a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
+index b20bfe5ab165bf86985e5ff2f93f415d9710e0e4..020632491b3ac5d78b7b91afe465c24459c3646a 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
+@@ -297,6 +297,28 @@ public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializa
+ 
+         return result;
+     }
++
++    @Override
++    public org.bukkit.advancement.AdvancementProgress getAdvancementProgress(org.bukkit.advancement.Advancement advancement) {
++        com.google.common.base.Preconditions.checkArgument(advancement != null, "advancement");
++
++        org.bukkit.craftbukkit.advancement.CraftAdvancement craft = (org.bukkit.craftbukkit.advancement.CraftAdvancement) advancement;
++        net.minecraft.server.PlayerAdvancements data = server.getHandle().getPlayerAdvancements(getUniqueId());
++        net.minecraft.advancements.AdvancementProgress progress = data.getOrStartProgress(craft.getHandle());
++        return new org.bukkit.craftbukkit.advancement.CraftAdvancementProgress(craft, data, progress) {
++            @Override
++            public boolean awardCriteria(String criteria) {
++                throw new UnsupportedOperationException("Cannot award criteria to an offline player");
```

Comment:
What if this is a `Player`?

Reply:
It's overridden in CraftPlayer right?

Reply:
I wouldn't know; add an `@Override` to both require that on compile, and to diff on change in that case

Reply:
No you are right I think, I can do a check to see if it can be a Player and then delegate to that.